components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/odata/MongoDataHandler.java

```diff
@@ -0,0 +1,427 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.dataservices.core.odata;
+
+import com.mongodb.DBCollection;
+import com.mongodb.DBObject;
+import com.mongodb.WriteResult;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.bson.Document;
+import org.bson.types.ObjectId;
+import org.jongo.Jongo;
+import org.json.JSONObject;
+import org.wso2.carbon.dataservices.core.description.query.MongoQuery;
+import org.wso2.carbon.dataservices.core.engine.DataEntry;
+
+/**
+ * This class implements MongoDB datasource related operations for ODataDataHandler.
+ */
+public class MongoDataHandler implements ODataDataHandler {
+
+    /**
+     * configuration ID is the ID given for the data service, at the time
+     * when the particular service is created.
+     */
+
+    private final String configId;
+
+    /**
+     * ObjectId s of the Collections
+     */
+    private Map<String, List<String>> primaryKeys;
+
+    /**
+     * List of Collections in the Database.
+     */
+    private List<String> tableList;
+
+    /**
+     * Metadata of the Collections
+     */
+    private Map<String, Map<String, DataColumn>> tableMetaData;
+
+    private Jongo jongo;
+    private static final String ETAG = "ETag";
+    private static final String DOCUMENT_ID = "_id";
+    private static final String SET = "{$set: {";
+
+    private ThreadLocal<Boolean> transactionAvailable = new ThreadLocal<Boolean>() {
+        protected synchronized Boolean initialValue() {
+
+            return false;
+        }
+    };
+
+    public MongoDataHandler(String configId, Jongo jongo) {
+
+        this.configId = configId;
+        this.jongo = jongo;
+        this.tableList = generateTableList();
+        this.tableMetaData = generateTableMetaData();
+        this.primaryKeys = generatePrimaryKeys();
+    }
+
+    /**
+     * This method returns database collection metadata.
+     * Returns a map with collection name as the key, and the values containing
+     * maps with column name as the map key, and the values of the column name
+     * map will be a DataColumn object, which represents the column.
+     *
+     * @return Database Metadata
+     * @see org.wso2.carbon.dataservices.core.odata.DataColumn
+     */
+    @Override
+    public Map<String, Map<String, DataColumn>> getTableMetadata() {
+
+        return this.tableMetaData;
+    }
+
+    private Map<String, Map<String, DataColumn>> generateTableMetaData() {
+
+        int i = 1;
+        Map<String, Map<String, DataColumn>> metaData = new HashMap<>();
+        HashMap<String, DataColumn> column = new HashMap<>();
+        for (String tableName : this.tableList) {
+            DBCollection readResult = jongo.getDatabase().getCollection(tableName);
+            Iterator<DBObject> cursor = readResult.find();
```

Comment:
Can readResult null? What happen if the table name is not exist?

Reply:
No. When a request is sent with a non-existing table name, before hitting this class, it recognizes that the particular table does not exist in the database during resource path analysis.